module/grpcserver/server.go

```diff
@@ -0,0 +1,176 @@
+	// was specified)
+	g.addrLock.Lock()
+	g.grpcAddress = l.Addr()
+	g.addrLock.Unlock()
+	g.log.Debug().Str("grpc_address", g.grpcAddress.String()).Msg("listening on port")
+	ready()
+
+	err = g.grpcServer.Serve(l) // blocking call
+	if err != nil {
+		g.log.Err(err).Msg("fatal error in grpc server")
+		ctx.Throw(err)
+	}
+}
+
+// GRPCAddress returns the listen address of the GRPC server.
+// Guaranteed to be non-nil after Engine.Ready is closed.
+func (g *GrpcServer) GRPCAddress() net.Addr {
+	g.addrLock.RLock()
+	defer g.addrLock.RUnlock()
+	return g.grpcAddress
+}
+
+// shutdownWorker is a worker routine which shuts down server when the context is cancelled.
+func (g *GrpcServer) shutdownWorker(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+	<-ctx.Done()
+	g.grpcServer.GracefulStop()
+}
+
+type GrpcServerBuilder struct {
+	log    zerolog.Logger
+	config GrpcServerConfig
+	server *grpc.Server
+}
+
+// NewGrpcServerBuilder helps to build a new grpc server.
+func NewGrpcServerBuilder(log zerolog.Logger,
+	config GrpcServerConfig,
+	rpcMetricsEnabled bool,
+	apiRateLimits map[string]int, // the api rate limit (max calls per second) for each of the Access API e.g. Ping->100, GetTransaction->300
+	apiBurstLimits map[string]int, // the api burst limit (max calls at the same time) for each of the Access API e.g. Ping->50, GetTransaction->10
+) *GrpcServerBuilder {
+	log = log.With().Str("component", "grpc_server").Logger()
+	// create a GRPC server to serve GRPC clients
+	grpcOpts := []grpc.ServerOption{
+		grpc.MaxRecvMsgSize(int(config.MaxMsgSize)),
+		grpc.MaxSendMsgSize(int(config.MaxMsgSize)),
+	}
+	var interceptors []grpc.UnaryServerInterceptor // ordered list of interceptors
+	// if rpc metrics is enabled, first create the grpc metrics interceptor
+	if rpcMetricsEnabled {
+		interceptors = append(interceptors, grpc_prometheus.UnaryServerInterceptor)
+	}
+	if len(apiRateLimits) > 0 {
+		// create a rate limit interceptor
+		rateLimitInterceptor := rpc.NewRateLimiterInterceptor(log, apiRateLimits, apiBurstLimits).UnaryServerInterceptor
+		// append the rate limit interceptor to the list of interceptors
+		interceptors = append(interceptors, rateLimitInterceptor)
+	}
+	// add the logging interceptor, ensure it is innermost wrapper
+	interceptors = append(interceptors, rpc.LoggingInterceptor(log)...)
+	// create a chained unary interceptor
+	chainedInterceptors := grpc.ChainUnaryInterceptor(interceptors...)
+	// create an unsecured grpc server
+	grpcOpts = append(grpcOpts, chainedInterceptors)
+	if config.TransportCredentials != nil {
+		// create a secure server by using the secure grpc credentials that are passed in as part of config
+		grpcOpts = append(grpcOpts, grpc.Creds(config.TransportCredentials))
+	}
+
+	return &GrpcServerBuilder{
+		log:    log,
+		config: config,
+		server: grpc.NewServer(grpcOpts...),
```

Comment:
In general, the main idea behind the creation of the 'GrpcServerBuilder' is to separate the process of creating and starting a gRPC server. This is necessary because services need to be registered before the server starts.

From the [RegisterService](https://pkg.go.dev/google.golang.org/grpc#ServiceRegistrar) doc

> // RegisterService registers a service and its implementation to the
> // concrete type implementing this interface. **It may not be called**
> // **once the server has started serving.**

### Idea diagram
`builder creates grpc server -> builder provides grpc server to register services -> builder starts grpc server`

That is the reason why I operated and saved GrpcServerBuilder instanses instead of GrpcServer instances, 
used Module("creating grpc servers"..) which is before builder.Component("RPC engine"...) and builder.Component("exec state stream engine"...). Also it is the reason why I moved serveGRPCWorkers from state_stream and rpc engines to GrpcServerBuilder->Build and added builder.Component("secure grpc server",...) and builder.Component("unsecure grpc server",...) after all service registered.

Reply:
We use the builder pattern quite a bit in this repo, and I agree it makes sense for this usecase. There's a bit of nuance around how the node bootstrapping works, which is where I'm coming from with this line of comments.

There are 2 levels of abstraction at play:
1. The individual component which manages _how_ it is started and stopped
2. The node builder which manages _when_ the components are started and stopped

First to provide a bit of context

When the node builder runs, it collects callbacks to run after all of the initialization is complete. All of the callbacks passed via `Module()` are run first, then all of those passed via `Component()`. These are called in the order they were added.

When the callbacks passed to `Component()` are run, the `Start()` method is called on the object that's returned. It then waits until the `Ready()` channel closes before moving on to the next one.

A `component.ComponentManager`, like what you used in `GrpcServer`, has a set of workers that it adds with `AddWorker()`. When `Start()` is called, all of the workers are then called. the `ComponentManager` is "ready" after all of the workers call their `ready()` callback.

So the order code is executed is:
1. All the code at the scope where calls to `builder.Component()` are run.
2. Each of the `Module()` callbacks in order
3. Each of the `Component()` callbacks, plus `Start()` on the returned object in order.

Now, in the case of `GrpcServer`, I believe the actual grpc server starts serving here:
https://github.com/onflow/flow-go/pull/4411/files#diff-78a699a0522765030d5043e4bf31ecefcef95ff2e0100c06bf6c0a27d8a7bfd3R65
which is run when `Start()` is called on the `GrpcServer` instance. This means that the component that returns the `GrpcServer` instance should be called last.

However, the `GrpcServerBuilder` only creates the server instance. My suggestion is to call `Build()` in the `Module()`s where you create the builders. This will produce a `GrpcServer` instance. The actual grpc server will not start until `Start()` is called, so you can use that instance to register all of the APIs you need. Finally, after all of those APIs are added, add the `Component()` that returns the `GrpcServer`.